httpserver/middlewares/validator.go

```diff
@@ -0,0 +1,120 @@
+package middlewares
+
+import (
+	"context"
+	"errors"
+	"fx.prodigy9.co/config"
+	"fx.prodigy9.co/httpserver/render"
+	"fx.prodigy9.co/validation"
+	"github.com/ggicci/httpin"
+	"github.com/go-chi/chi/v5"
+	"net/http"
+	"reflect"
+)
+
+/**
+ * httpin package used to read request data into structs
+ * govalidator package used for validation
+ *
+ * Use header "x-api-action" for conditional validation
+ */
+
+// Validator adds a validator to the request
+func Validator(cfg *config.Source) func(http.Handler) http.Handler {
+	val := &validation.Validator{}
+	val.Init()
+	val.AddTranslations(validation.Translations{
+		"required":             "{0} is required",
+		"required_if":          "{0} is required",
+		"required_with":        "{0} is required",
+		"required_unless":      "{0} is required",
+		"required_without_all": "{0} is required",
+		"required_without":     "{0} is required",
+		"email":                "Email is invalid",
+	})
+
+	httpin.UseGochiURLParam("path", chi.URLParam)
+	httpin.RegisterDirectiveExecutor("validator", httpin.DirectiveExecutorFunc(executor(val)), nil)
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			ctx := context.WithValue(req.Context(), "validator", val)
+			next.ServeHTTP(rw, req.WithContext(ctx))
+		})
+	}
+}
+
+// executor will validate a struct field marked with the "validator" tag
+func executor(val *validation.Validator) func(ctx *httpin.DirectiveContext) error {
+	return func(ctx *httpin.DirectiveContext) error {
+		action := ctx.Request.Header.Get("x-api-action")
+		if field := ctx.Value.Elem().FieldByName("Action"); field != (reflect.Value{}) {
+			field.SetString(action)
+		}
+		return val.Validate(ctx.Value.Elem().Interface(), ctx.ValueType)
+	}
+}
+
+// errorHandler will unwrap httpin error to pass the fx validation errors to render
+func errorHandler(rw http.ResponseWriter, r *http.Request, err error) {
```

Comment:
I ran into a problem with this function.
when I make a request with something like this
```
{
   "name":"Finance",
   "code": "FIN" 
   // comment
}
```
It doesn't render error from requesToStruct